tools/building.py

```diff
@@ -932,6 +932,9 @@ def metadce(js_file, wasm_file, minify_whitespace, debug_info):
       export = asmjs_mangle(item['export'])
       if settings.EXPORT_ALL or export in required_symbols:
         item['root'] = True
+    if 'import' in item and item['import'][0] == 'env' and \
+        item['import'][1] == '___cpp_exception':
+      item['root'] = True
```

Comment:
Is this for the case where the main module does not use exceptions (so it does not use this import) but the side module does?

Makes sense if so, but please add a comment.

Reply:
Interesting. Where is `__cpp_exception` used from, say in the simple case of just a main module? If it's used from wasm, then metadce shouldn't be able to remove it (can build with `EMCC_DEBUG=1` and look at the wasm before and after metadce runs to confirm); if it's used from JS then this might make sense, but where is that use?

Reply:
Interesting. So the wasm needs to read that tag from JS, and it has an import statement for it. Then this sounds like a metadce bug, where it doesn't properly handle uses of tags.

The tool does have general tag handling,

https://github.com/WebAssembly/binaryen/blob/7126a43e9e1cea50d3534b50d4554f4a18cd6e14/src/tools/wasm-metadce.cpp#L181-L186

But what I don't see is logic to scan tags in function bodies - so it doesn't see that tag is used. Here is the code to scan calls for their targets:

https://github.com/WebAssembly/binaryen/blob/7126a43e9e1cea50d3534b50d4554f4a18cd6e14/src/tools/wasm-metadce.cpp#L254-L265

I think we need something similar for instructions that reference a tag?

With all that said, if this is urgent I'm ok to land this with a root of that one tag as a workaround for now. It's just one, so it's not a code size issue. If so, how about just adding a TODO here?